stdlib/public/runtime/ReflectionMirror.mm

```diff
@@ -0,0 +1,790 @@
+struct ReflectionMirrorImpl {
+  const Metadata *type;
+  OpaqueValue *value;
+  
+  virtual char displayStyle() = 0;
+  virtual intptr_t count() = 0;
+  virtual AnyReturn subscript(intptr_t index, const char **outName,
+                              void (**outFreeFunc)(const char *)) = 0;
+  virtual const char *enumCaseName() { return nullptr; }
+
+#if SWIFT_OBJC_INTEROP
+  virtual id quickLookObject() { return nil; }
+#endif
+  
+  virtual ~ReflectionMirrorImpl() {}
+};
+
+
+// Implementation for tuples.
+struct TupleImpl: ReflectionMirrorImpl {
+  char displayStyle() {
+    return 't';
+  }
+  
+  intptr_t count() {
+    auto Tuple = static_cast<const TupleTypeMetadata *>(type);
+    return Tuple->NumElements;
+  }
+  
+  AnyReturn subscript(intptr_t i, const char **outName,
+                      void (**outFreeFunc)(const char *)) {
+    auto Tuple = static_cast<const TupleTypeMetadata *>(type);
+
+    if (i < 0 || (size_t)i > Tuple->NumElements)
+      swift::crash("Swift mirror subscript bounds check failure");
+
+    // Determine whether there is a label.
+    bool hasLabel = false;
+    if (const char *labels = Tuple->Labels) {
+      const char *space = strchr(labels, ' ');
+      for (intptr_t j = 0; j != i && space; ++j) {
+        labels = space + 1;
+        space = strchr(labels, ' ');
+      }
+
+      // If we have a label, create it.
+      if (labels && space && labels != space) {
+        *outName = strndup(labels, space - labels);
+        hasLabel = true;
+      }
+    }
+
+    if (!hasLabel) {
+      // The name is the stringized element number '.0'.
+      char *str;
+      asprintf(&str, ".%" PRIdPTR, i);
+      *outName = str;
+    }
+    
+    *outFreeFunc = [](const char *str) { free(const_cast<char *>(str)); };
+
+    // Get the nth element.
+    auto &elt = Tuple->getElement(i);
+    auto bytes = reinterpret_cast<const char*>(value);
+    auto eltData = reinterpret_cast<const OpaqueValue *>(bytes + elt.Offset);
+
+    Any result;
+
+    result.Type = elt.Type;
+    auto *opaqueValueAddr = result.Type->allocateBoxForExistentialIn(&result.Buffer);
+    result.Type->vw_initializeWithCopy(opaqueValueAddr,
+                                       const_cast<OpaqueValue *>(eltData));
+
+    return AnyReturn(result);
+  }
+};
+
+
+// Implementation for structs.
+struct StructImpl: ReflectionMirrorImpl {
+  char displayStyle() {
+    return 's';
+  }
+  
+  intptr_t count() {
+    auto Struct = static_cast<const StructMetadata *>(type);
+    return Struct->Description->Struct.NumFields;
+  }
+  
+  AnyReturn subscript(intptr_t i, const char **outName,
+                      void (**outFreeFunc)(const char *)) {
+    auto Struct = static_cast<const StructMetadata *>(type);
+
+    if (i < 0 || (size_t)i > Struct->Description->Struct.NumFields)
+      swift::crash("Swift mirror subscript bounds check failure");
+
+    // Load the type and offset from their respective vectors.
+    auto fieldType = Struct->getFieldTypes()[i];
```

Comment:
Hmm, this is going to conflict with @xedin 's current work in https://github.com/apple/swift/pull/13926, which eliminates the field-type vector. There will be an accessor of some sort, so the conflict resolution should be simple either way.

Reply:
I think it shouldn't be too bad :)